### PR TITLE
Add seccomp and apparmor support.

### PR DESCRIPTION
**Pull request by Random-Liu**
This PR adds seccomp and apparmor support in new CRI.

This a WIP because I'm still adding unit test for some of the functions. Sent this PR here for design discussion.

This PR is similar with https://github.com/kubernetes/kubernetes/pull/33450.
The differences are:
- This PR passes seccomp and apparmor configuration via annotations;
- This PR keeps the seccomp handling logic in docker shim because current seccomp implementation is very docker specific, and @timstclair told me that even the json seccomp profile file is defined by docker.

Notice that this PR almost passes related annotations in `api.Pod` to the runtime directly instead of introducing new CRI annotation.

@yujuhong @feiskyer @timstclair

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33500)

### CODE REVIEW

**Comment by yifan-gu** [comment]
+1 on this direction than #33450.

cc @lucab @s-urbaniak @jonboulle

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go [hunk 1]
```diff
@@ -48,6 +48,16 @@ func (m *kubeGenericRuntimeManager) createPodSandbox(pod *api.Pod, attempt uint3
 	return podSandBoxID, "", nil
 }
 
+// applyPodConfigAnnotations applies some pod configurations into annotations. These configurations
+// are usually experimental features. They'll be promoted to official runtime api in the future.
+func applySandboxConfigAnnotations(annotations map[string]string, pod *api.Pod) {
```
Comment (feiskyer) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox.go]: We don't need this actually, `newPodAnnotations` returns exactly `pod.Annotations`.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 2]
```diff
@@ -117,6 +118,25 @@ func getContainerLogsPath(containerName string, podUID types.UID) string {
 	return path.Join(podLogsRootDirectory, string(podUID), fmt.Sprintf("%s.log", containerName))
 }
 
+// applyContainerConfigAnnotations applies some container configurations into annotations. These configurations
+// are usually experimental features. They'll be promoted to official runtime api in the future.
+func applyContainerConfigAnnotations(annotations map[string]string, pod *api.Pod, container *api.Container) {
```
Comment (feiskyer) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: Same here, those annotations are already in `SandboxConfig.Annotations`.

**Comment by feiskyer** [comment]
@Random-Liu It's better to also add documents about app armor and seccomp in CRI (e.g. on `SandboxConfig.Annotations`) .

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox.go]: I see. If we assume `pod.Annotations` should be passed to runtime, we can remove this.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: It seems that we are not passing `pod.Annotations` to container, so we may still need this function.

Comment (feiskyer) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: It's in the sandbox config, which is also a param of CreateContainer.

**Comment by Random-Liu** [comment]
@feiskyer @yujuhong I assume that we'll have a CRI API document in the future right? If so, we can document annotation based api there. :)

**Comment by feiskyer** [comment]
> @feiskyer @yujuhong I assume that we'll have a CRI API document in the future right? If so, we can document annotation based api there. :)

Yes, doc should be added before release.

> One question is that we should we use annotation api? Only for Kubernetes alpha/beta features?

I'm +1 for this.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: Ha, yeah, you are right. Then we are good, just give it to dockershim. Things become even easier!

**Comment by yujuhong** [comment]
> @feiskyer @yujuhong I assume that we'll have a CRI API document in the future right? If so, we can document annotation based api there. :)

Yes, we should definitely have more documents.

**Comment by yujuhong** [comment]
Reviewed 7 of 8 files at r1.
Review status: 7 of 8 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/kubelet.go, line 553 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSgkNczzOl4C_dnQDbi:-KSgkNczzOl4C_dnQDbj:b-2q9h7k) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/kubelet.go#L553)):*
> ```Go
> 			// Use the new CRI shim for docker. This is need for testing the
> 			// docker integration through CRI, and may be removed in the future.
> 			dockerService := dockershim.NewDockerService(klet.dockerClient, kubeCfg.SeccompProfileRoot)
> ```

Just a note: anything we passed to DockerService may eventually need to handled in another way if/when we switch to a running the shim as a different process.

---

*[pkg/kubelet/dockershim/docker_container.go, line 81 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSgld-hL2avZ9GaEf_5:-KSgld-hL2avZ9GaEf_6:bclbi20) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/docker_container.go#L81)):*
> ```Go
> 
> // getContainerSecurityOpts gets security options from container config. (seccomp and apparmor)
> func (ds *dockerService) getContainerSecurityOpts(config *runtimeApi.ContainerConfig) ([]string, error) {
> ```

nit: Since the function doesn't use much from *dockerService, how about making this a standalone function for easy testing?

---

*[pkg/kubelet/dockershim/docker_container.go, line 85 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSgmE9FW8qxqjfS458m:-KSgmE9FW8qxqjfS458n:b-o4x86u) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/docker_container.go#L85)):*
> ```Go
> 	if seccompOpts, err := generateContainerSeccompOpts(config, ds.seccompProfileRoot); err != nil {
> 		return nil, err
> 	} else {
> ```

`else` is not needed.

---

*[pkg/kubelet/dockershim/docker_container.go, line 90 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSgmHXZLUnli3_83dIX:-KSgmHXZLUnli3_83dIY:bxla6hw) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/docker_container.go#L90)):*
> ```Go
> 	if appArmorOpts, err := generateContainerAppArmorOpts(config); err != nil {
> 		return nil, err
> 	} else {
> ```

`else` is not needed

---

*[pkg/kubelet/dockershim/docker_sandbox.go, line 205 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSgmZW9gHa-KcCiA2jW:-KSgmZW9gHa-KcCiA2jX:b4suqxk) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/docker_sandbox.go#L205)):*
> ```Go
> 	if seccompOpts, err := generateSandboxSeccompOpts(config, ds.seccompProfileRoot); err != nil {
> 		return nil, err
> 	} else {
> ```

`else` not needed.

---

*[pkg/kubelet/dockershim/docker_service.go, line 81 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSgmlL7fdt9m3g3YLEx:-KSgmlL7fdt9m3g3YLEy:b-rrhaq0) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/docker_service.go#L81)):*
> ```Go
> type dockerService struct {
> 	// TODO: Current seccomp implementation is very docker specific. Move this somewhere else
> 	// after we define more general seccomp api.
> ```

nit: It's unclear where we will move this to.

s/define more/define a more

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by timstclair** [comment]
Review status: 4 of 8 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/dockershim/helpers.go, line 195 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSh9ZN6OUvOBqjq4GbJ:-KSh9ZN6OUvOBqjq4GbK:bmz57fe) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/helpers.go#L195)):*
> ```Go
> 	// Assume validation has already happened.
> 	profileName := strings.TrimPrefix(profile, apparmor.ProfileNamePrefix)
> 	return []string{fmt.Sprintf("apparmor=%s", profileName)}, nil
> ```

Are  we committing to only supporting Docker >= 1.11 for this? Otherwise this needs tho version-dependent formatting defined [here](https://github.com/kubernetes/kubernetes/pull/33500)

---

*[pkg/kubelet/dockershim/helpers.go, line 201 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KShA3ysRVVj_n8twuIJ:-KShA3ytgwgMTWbFpt1p:bb067yo) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/helpers.go#L201)):*
> ```Go
> func generateContainerSeccompOpts(config *runtimeApi.ContainerConfig, seccompProfileRoot string) ([]string, error) {
> 	annotations := config.GetAnnotations()
> 	profile := annotations[api.SeccompContainerAnnotationKeyPrefix+config.Metadata.GetName()]
> ```

If the profile is not set this should use the pod profile.

---

*[pkg/kubelet/dockershim/helpers.go, line 212 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KShADq9jOJ0cMryq4dd:-KShADq9jOJ0cMryq4de:b7cifwp) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/helpers.go#L212)):*
> ```Go
> }
> 
> func generateSeccompOpts(profile, seccompProfileRoot string) ([]string, error) {
> ```

It makes me nervous that we're forking all this logic. Can we extract higher-level annotations to use both here and in dockertool? I believe the important pieces of input are the docker version, and the annotations. From there, we can get an array of formatted []string opts, and an error.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
*[pkg/kubelet/kubelet.go, line 553 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSgkNczzOl4C_dnQDbi:-KShPUtJxsyhI1liFSKG:b-z7kihd) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/kubelet.go#L553)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Just a note: anything we passed to DockerService may eventually need to handled in another way if/when we switch to a running the shim as a different process.
</details>
ACK. Will add a note in the code.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
*[pkg/kubelet/dockershim/docker_container.go, line 81 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSgld-hL2avZ9GaEf_5:-KShPcgfj7WfS1plnuRk:b2nha9w) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/docker_container.go#L81)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> nit: Since the function doesn't use much from *dockerService, how about making this a standalone function for easy testing?
</details>
ACK

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
*[pkg/kubelet/dockershim/docker_container.go, line 85 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSgmE9FW8qxqjfS458m:-KShPj3uM7wXIBjOIOjD:b-hhs3iu) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/docker_container.go#L85)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> `else` is not needed.
</details>
Copied from original code, will remove it

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
*[pkg/kubelet/dockershim/helpers.go, line 195 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSh9ZN6OUvOBqjq4GbJ:-KShPsuVB3kLDWvc5Wje:b-2kmqbf) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/helpers.go#L195)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Are  we committing to only supporting Docker >= 1.11 for this? Otherwise this needs tho version-dependent formatting defined [here](https://github.com/kubernetes/kubernetes/pull/33500)
</details>
Yeah, currently only Docker >= 1.11, I'm also tracking the compatibility issue [here](https://github.com/kubernetes/kubernetes/issues/29478#issuecomment-248197665)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
*[pkg/kubelet/dockershim/helpers.go, line 201 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KShA3ysRVVj_n8twuIJ:-KShQ2Kib2dcNmo0kldu:bud7aod) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/helpers.go#L201)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> If the profile is not set this should use the pod profile.
</details>
Done in the kubelet level. However, the logic will change a bit because @feiskyer told me that Annotations are already passed to runtime by default, will add the logic here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
*[pkg/kubelet/dockershim/helpers.go, line 212 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KShADq9jOJ0cMryq4dd:-KShQdD0lFflrmaIsge4:b-wuyr3j) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/helpers.go#L212)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> It makes me nervous that we're forking all this logic. Can we extract higher-level annotations to use both here and in dockertool? I believe the important pieces of input are the docker version, and the annotations. From there, we can get an array of formatted []string opts, and an error.
</details>
We've already reused some logic from dockertools - `dockertools.GetSeccompOpts`. I'll change the PR to also share the annotation related logic.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by yujuhong** [comment]
Review status: 4 of 8 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/dockershim/helpers.go, line 195 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSh9ZN6OUvOBqjq4GbJ:-KShgvTsspLFRR_O-ZzB:b9dv36a) ([raw file](https://github.com/kubernetes/kubernetes/blob/8f86ea544bd43cc936f6da2ef29a6b90101d5ccf/pkg/kubelet/dockershim/helpers.go#L195)):*
<details><summary><i>Previously, Random-Liu (Lantao Liu) wrote…</i></summary>
> Yeah, currently only Docker >= 1.11, I'm also tracking the compatibility issue [here](https://github.com/kubernetes/kubernetes/issues/29478#issuecomment-248197665)
</details>
We will support docker 1.11+ for sure. We won't support docker 1.9 and below. As for docker 1.10, it's TBD and I am leaning towards to not supporting it. We'll add address 1.10 compatibility issues if we decide to support it. 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
I've run seccomp and apparmor test with cri enabled, all tests passed.
I'll send another PR to migrate seccomp test to node e2e.

**Comment by timstclair** [comment]
Review status: 0 of 11 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/security/apparmor/helpers.go, line 51 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSmQ6QIXAwK1IsE8Pbc:-KSmQ6QIXAwK1IsE8Pbd:b-5e5r2y) ([raw file](https://github.com/kubernetes/kubernetes/blob/858390f0f1b1f66cad400ffe1048aa1e759c4968/pkg/security/apparmor/helpers.go#L51)):*
> ```Go
> 
> // Returns the name of the profile to use with the container.
> func GetProfileName(pod *api.Pod, containerName string) string {
> ```

In this method used any more?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
*[pkg/security/apparmor/helpers.go, line 51 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSmQ6QIXAwK1IsE8Pbc:-KSmZeibX8Tj9t8eE8aw:buh36md) ([raw file](https://github.com/kubernetes/kubernetes/blob/858390f0f1b1f66cad400ffe1048aa1e759c4968/pkg/security/apparmor/helpers.go#L51)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> In this method used any more?
</details>
Nope. Are you ok with removing it?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by timstclair** [comment]
Review status: 0 of 11 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/security/apparmor/helpers.go, line 51 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSmQ6QIXAwK1IsE8Pbc:-KSma52TNslMBL-grRlZ:b-ybnhhd) ([raw file](https://github.com/kubernetes/kubernetes/blob/858390f0f1b1f66cad400ffe1048aa1e759c4968/pkg/security/apparmor/helpers.go#L51)):*
<details><summary><i>Previously, Random-Liu (Lantao Liu) wrote…</i></summary>
> Nope. Are you ok with removing it?
</details>
Yeah. Maybe rename `GetProfileNameFromPodAnnotations` to `GetProfileName`?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
*[pkg/security/apparmor/helpers.go, line 51 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSmQ6QIXAwK1IsE8Pbc:-KSmaoI12KjGNnGBLLiK:bhy9ot6) ([raw file](https://github.com/kubernetes/kubernetes/blob/858390f0f1b1f66cad400ffe1048aa1e759c4968/pkg/security/apparmor/helpers.go#L51)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Yeah. Maybe rename `GetProfileNameFromPodAnnotations` to `GetProfileName`?
</details>
Will do.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

**Comment by Random-Liu** [comment]
*[pkg/security/apparmor/helpers.go, line 51 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500#-KSmQ6QIXAwK1IsE8Pbc:-KSmk_Dde_Jx_0YSjf4Y:bffluw4) ([raw file](https://github.com/kubernetes/kubernetes/blob/858390f0f1b1f66cad400ffe1048aa1e759c4968/pkg/security/apparmor/helpers.go#L51)):*
<details><summary><i>Previously, Random-Liu (Lantao Liu) wrote…</i></summary>
> Will do.
</details>
I find that this function is used in the other apparmor related packages validation, PSP etc. We can change all of them, but maybe we'd better keep the original function.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33500)*

File: pkg/kubelet/dockershim/helpers_test.go [hunk 3]
```diff
@@ -32,3 +36,114 @@ func TestLabelsAndAnnotationsRoundTrip(t *testing.T) {
 	assert.Equal(t, expectedLabels, actualLabels)
 	assert.Equal(t, expectedAnnotations, actualAnnotations)
 }
+
+// TestGetContainerSecurityOpts tests the logic of generating container security options from sandbox annotations.
+// The actual profile loading logic is tested in dockertools.
+// TODO: Migrate the corresponding test to dockershim.
```
Comment (yujuhong) [inline pkg/kubelet/dockershim/helpers_test.go]: nit: migrate to dockershim from where?

**Comment by feiskyer** [comment]
LGTM

**Comment by yujuhong** [comment]
LGTM. Leaving to @timstclair for the final round.

**Comment by timstclair** [comment]
LGTM

**Comment by Random-Liu** [comment]
@k8s-bot test this issue #IGNORE.

**Pull request merged**